sound/soc/sof/virtio-fe.c

```diff
@@ -0,0 +1,835 @@
+// SPDX-License-Identifier: (GPL-2.0 OR BSD-3-Clause)
+/*
+ * This file is provided under a dual BSD/GPLv2 license.  When using or
+ * redistributing this file, you may do so under either license.
+ *
+ * Copyright(c) 2017-2019 Intel Corporation. All rights reserved.
+ *
+ * Author:	Libin Yang <libin.yang@intel.com>
+ *		Luo Xionghu <xionghu.luo@intel.com>
+ *		Liam Girdwood <liam.r.girdwood@linux.intel.com>
+ *		Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>
+ */
+
+/*
+ * virt IO FE driver
+ *
+ * The SOF driver thinks this driver is another audio DSP, however the calls
+ * made by the SOF driver core do not directly go to HW, but over a virtIO
+ * message Q to the virtIO BE driver.
+ *
+ * The virtIO message Q uses the *exact* same IPC structures as we currently
+ * use in the mailbox.
+ *
+ * The mailbox IO and TX/RX msg functions below do IO on the virt IO Q.
+ */
+
+#include <linux/completion.h>
+#include <linux/device.h>
+#include <linux/firmware.h>
+#include <linux/interrupt.h>
+#include <linux/module.h>
+#include <linux/scatterlist.h>
+#include <linux/virtio.h>
+#include <linux/virtio_config.h>
+#include <linux/virtio_ids.h>
+#include <linux/virtio_ring.h>
+#include <sound/sof.h>
+#include <sound/sof/virtio.h>
+
+#include "ops.h"
+#include "sof-audio.h"
+#include "sof-priv.h"
+
+#define SOF_COMPONENT_NAME	"sof-audio-component"
+/* 600ms for VirtQ IPC */
+#define DATA_TIMEOUT_MS		1000
+
+static const char *const sof_vq_names[SOF_VIRTIO_NUM_OF_VQS] = {
+	SOF_VIRTIO_IPC_CMD_VQ_NAME,
+	SOF_VIRTIO_IPC_PSN_VQ_NAME,
+	SOF_VIRTIO_DATA_VQ_NAME,
+};
+
+struct sof_vfe {
+	struct snd_sof_dev *sdev;
+
+	/* IPC cmd from frontend to backend */
+	struct virtqueue *ipc_cmd_vq;
+
+	/* IPC position update from backend to frontend */
+	struct virtqueue *ipc_psn_vq;
+
+	/* audio data in both directions */
+	struct virtqueue *data_vq;
```

Comment:
@lyakh what determines the fact that the first 2 vq's are one-way and the third one is bidirectional?

Reply:
@ranj063 they all are bidirectional. The comments only explain in which direction the payload is travelling, but packets are sent in both directions in all of them. In the IPC case IPCs are sent to the host, then the status is returned. In the position update case, position data is sent from the host to the guest, but the guest has to first send a message with a buffer for the host to put that position data into. So, they all are bidirectional